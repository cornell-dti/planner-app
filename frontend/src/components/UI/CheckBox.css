.CheckBox {
  display: inline-block;
  position: relative;
  height: 16px;
  width: 16px;
  min-width: 16px;
  cursor: pointer;
  border: 2px solid white;
  user-select: none;
}

.InvertedCheckBox {
  border-color: #333333;
}

.CheckBox input {
  position: absolute;
  cursor: pointer;
  opacity: 0;
  height: 0;
  width: 0;
}

.CheckBoxCheckMark {
  position: absolute;
  cursor: pointer;
  top: 0;
  left: 0;
  height: 12px;
  width: 12px;
  background-color: transparent;
}

.CheckBox:hover input ~ .CheckBoxCheckMark {
  background-color: rgba(255, 255, 255, 0.15);
}

.CheckBox input:checked ~ .CheckBoxCheckMark {
  background-color: rgba(255, 255, 255, 0.15);
}

.CheckBoxCheckMark:after {
  position: absolute;
  content: "";
}

.CheckBox .CheckBoxCheckMark:after {
  left: 4px;
  top: 0;
  width: 5px;
  height: 10px;
  border: solid white;
  border-width: 0 2px 2px 0;
  transform: rotate(45deg);
}

.InvertedCheckBox .CheckBoxCheckMark:after {
  border-color: #333333;
}
