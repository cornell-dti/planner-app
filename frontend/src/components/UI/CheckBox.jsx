// @flow strict
/* eslint-disable jsx-a11y/label-has-associated-control,jsx-a11y/label-has-for */

import React from 'react';
import type { Node } from 'react';
import styles from './CheckBox.css';

type Props = {|
  className?: string; // additional className to apply
  checked?: boolean; // whether the box is initially checked
  inverted?: boolean; // whether the color is inverted.
  +onChange: (checked: boolean) => any; // called when the value changed.
|};

type State = {| +checked: boolean |};

/**
 * This is the checkbox that implements the minimalist design.
 */
export default class CheckBox extends React.PureComponent<Props, State> {
  static defaultProps = {
    className: '',
    checked: false,
    inverted: false,
  };

  constructor(props: Props) {
    super(props);
    const { checked } = props;
    this.state = { checked: checked || false };
  }

  componentDidUpdate({ checked }: Props, prevState: State) {
    if (checked !== prevState.checked) {
      // I have to disable this lint because I don't know any other way to solve it.
      // eslint-disable-next-line react/no-did-update-set-state
      this.setState({ checked });
    }
  }

  /**
   * Handle the click of the checkbox.
   */
  handleClick() {
    const { onChange } = this.props;
    this.setState(({ checked }: State) => {
      const newChecked: boolean = !checked;
      if (checked !== newChecked) {
        onChange(newChecked);
      }
      return { checked: newChecked };
    });
  }

  render(): Node {
    const { className, inverted } = this.props;
    let { checked } = this.state;
    // eslint-disable-next-line react/destructuring-assignment
    const checkedNotNull = this.props.checked || false;
    const invertedNotNull = inverted || false;
    if (checkedNotNull !== checked) {
      checked = checkedNotNull;
    }
    let allClassNames = className != null ? `${className} ${styles.CheckBox}` : styles.CheckBox;
    if (invertedNotNull) {
      allClassNames = `${allClassNames} ${styles.InvertedCheckBox}`;
    }
    const checkMarkDisplayStyle = { display: checked ? 'block' : 'none' };
    return (
      <label className={allClassNames}>
        <input
          tabIndex={-1}
          defaultChecked={checked}
          onClick={() => this.handleClick()}
          type="checkbox"
        />
        <span className={styles.CheckBoxCheckMark} style={checkMarkDisplayStyle} />
      </label>
    );
  }
}
