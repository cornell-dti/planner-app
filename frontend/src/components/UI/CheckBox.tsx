/* eslint-disable jsx-a11y/label-has-associated-control,jsx-a11y/label-has-for */

import React, { MouseEvent, ReactElement } from 'react';
import styles from './CheckBox.css';

type Props = {
  readonly checked: boolean; // whether the box is initially checked
  readonly onChange: (checked: boolean) => void; // called when the value changed.
  readonly disabled: boolean; // whether the checkbox is disabled.
  readonly inverted: boolean; // whether the color is inverted.
  readonly className: string | null; // additional className to apply
};

/**
 * This is the checkbox that implements designers' minimalist design.
 */
export default function CheckBox(props: Props): ReactElement {
  const {
    checked, onChange, disabled, inverted, className,
  } = props;
  let allClassNames = className === null ? styles.CheckBox : `${className} ${styles.CheckBox}`;
  if (inverted) {
    allClassNames = `${allClassNames} ${styles.InvertedCheckBox}`;
  }
  const handleClick = (e: MouseEvent<HTMLInputElement>): void => {
    e.stopPropagation();
    if (!disabled) {
      onChange(!checked);
    }
  };
  return (
    <label className={allClassNames}>
      <input
        tabIndex={-1}
        defaultChecked={checked}
        onClick={handleClick}
        disabled={disabled}
        type="checkbox"
      />
      {checked && (<span className={styles.CheckBoxCheckMark} />)}
    </label>
  );
}

CheckBox.defaultProps = { disabled: false, inverted: false, className: null };
