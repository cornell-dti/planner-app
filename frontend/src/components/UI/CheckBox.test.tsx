import React from 'react';
import { renderIntoDocument } from 'react-dom/test-utils';
import CheckBox from './CheckBox';

it('CheckBox can render', () => {
  const onChange = (): void => { };
  renderIntoDocument(<CheckBox checked onChange={onChange} />);
  renderIntoDocument(<CheckBox checked={false} onChange={onChange} />);
  renderIntoDocument(<CheckBox checked onChange={onChange} inverted={false} />);
  renderIntoDocument(<CheckBox checked={false} onChange={onChange} inverted={false} />);
  renderIntoDocument(<CheckBox checked onChange={onChange} inverted />);
  renderIntoDocument(<CheckBox checked={false} onChange={onChange} inverted />);
  renderIntoDocument(<CheckBox checked onChange={onChange} disabled={false} />);
  renderIntoDocument(<CheckBox checked={false} onChange={onChange} disabled={false} />);
  renderIntoDocument(<CheckBox checked onChange={onChange} disabled />);
  renderIntoDocument(<CheckBox checked={false} onChange={onChange} disabled />);
});
