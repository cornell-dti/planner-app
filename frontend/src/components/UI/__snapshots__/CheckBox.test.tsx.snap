// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CheckBox matches snapshot. 1`] = `
<label
  className="CheckBox Enabled"
>
  <input
    defaultChecked={true}
    disabled={false}
    onClick={[Function]}
    tabIndex={0}
    type="checkbox"
  />
  <span
    className="CheckBoxCheckMark"
  />
</label>
`;
